README.md
# teamspeak-query-client

[![Greenkeeper badge](https://badges.greenkeeper.io/mkg20001/teamspeak-query-client.svg)](https://greenkeeper.io/)

[![Build Status](https://travis-ci.org/mkg20001/teamspeak-query-client.svg?branch=master)](https://travis-ci.org/mkg20001/teamspeak-query-client)
[![codecov](https://codecov.io/gh/mkg20001/teamspeak-query-client/branch/master/graph/badge.svg)](https://codecov.io/gh/mkg20001/teamspeak-query-client)

A TeamSpeak Query Client that works.

# Example

```js
const query = new Query({ //you can also leave this blank, then it connects to localhost:10011
  host: "yourserver.de",
  port: 10011
})
query.connect(err => {
  if (err) throw err //Likely connection refused
  query.login("serveradmin", "mysecretpw", err => { //you can also skip login to opperate as guest query
    if (err && err.id === 520) throw new Error("Invalid Password")
    if (err) throw err //something weird happened
    query.cmd("use", 1, err => { //this is easy and works.
      if (err) throw err //likely no permissions
      query.list("clientlist", ["-uid"], (err, res) => { // "list" ensures res is an array at all time
        if (err) throw err //likely no permissions
        console.log(res)
      })
    })

    query.cmd("servernotifyregister", { event: "server" }, err => { //subscribe to "server" events
      if (err) {
        throw err
      } else {
        query.on("cliententerview", console.log) //user has joined
        query.on("clientleftview", console.log) //user has left
        query.on("notify", console.log) //all the events
      }
    })
  })
})
```

## There are hundreds of ts3 clients. Why write yet another one?

I've seen many ts3 querys for nodeJS. Every client had it's own flaws.

Some didn't even really work, some had trouble dealing with more than one server.

So I've written mine in the hope it works


CHANGELOG.md
<a name="0.1.6"></a>
## [0.1.6](https://github.com/mkg20001/teamspeak-query-client/compare/v0.1.5...v0.1.6) (2018-05-01)


### Features

* Fully rewrite without pull-queue - codestyle ([d4b253b](https://github.com/mkg20001/teamspeak-query-client/commit/d4b253b))



<a name="0.1.5"></a>
## [0.1.5](https://github.com/mkg20001/teamspeak-query-client/compare/v0.1.4...v0.1.5) (2018-01-03)


### Features

* Cleanup ([06df630](https://github.com/mkg20001/teamspeak-query-client/commit/06df630))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/mkg20001/teamspeak-query-client/compare/v0.1.3...v0.1.4) (2017-07-31)



<a name="0.1.3"></a>
## [0.1.3](https://github.com/mkg20001/teamspeak-query-client/compare/v0.1.2...v0.1.3) (2017-07-30)



<a name="0.1.2"></a>
## [0.1.2](https://github.com/mkg20001/teamspeak-query-client/compare/v0.1.1...v0.1.2) (2017-07-30)



<a name="0.1.1"></a>
## [0.1.1](https://github.com/mkg20001/teamspeak-query-client/compare/v0.1.0...v0.1.1) (2017-07-30)



<a name="0.1.0"></a>
# [0.1.0](https://github.com/mkg20001/teamspeak-query-client/compare/v0.0.1...v0.1.0) (2017-07-30)



<a name="0.0.1"></a>
## 0.0.1 (2017-07-30)





test/node.js
'use strict'

/* eslint-disable */

const chai = require('chai')
chai.use(require('sinon-chai'))
chai.should()
const expect = chai.expect
const assert = require('assert')

const Query = require('../src')
const mod = require('../src/pull')
const pull = require('pull-stream')
const FakeServer = require('./fake-ts3')

describe('query', () => {
  describe('connection', () => {
    it('should connect and login', next => {
      let fake = new FakeServer({
        login: ['serveradmin', 'pw123']
      })
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.login('serveradmin', 'pw123', err => err ? next(err) : true)
      q.disconnect()
    })

    it('should login, set the nickname and get the clientlist', next => {
      let fake = new FakeServer({
        login: ['serveradmin', 'pw123']
      })
      let cllist = [{
        args: {
          clid: 1,
          cid: 1,
          client_database_id: 1,
          client_nickname: 'Some Guy'
        }
      },
        {
          args: {
            clid: 2,
            cid: 2,
            client_database_id: 2,
            client_nickname: 'Another Guy'
          }
        }
      ]
      fake.addTarget('clientupdate')
      fake.addTarget('use')
      fake.addTarget('clientlist', cllist)
      fake.addTarget('clientlist', cllist)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.login('serveradmin', 'pw123', err => err ? next(err) : true)
      q.cmd('clientupdate', {
        client_nickname: 'test'
      }, err => err ? next(err) : true)
      q.cmd('use', 1, err => err ? next(err) : true)
      q.cmd('clientlist', (err, res) => err ? next(err) : assert.deepEqual(cllist.map(c => c.args), res, 'list differs'))
      q.list('clientlist', ['-uid'], (err, res) => err ? next(err) : assert.deepEqual(cllist.map(c => c.args), res, 'list differs'))
      q.disconnect()
    })

    it('should handle notify events', next => {
      let fake = new FakeServer({})
      let testdata = {
        bools: ['notifyclientleftview'],
        args: {
          client_nickname: 'Another guy'
        }
      }
      fake.addTarget('servernotifyregister', testdata)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.on('clientleftview', res => {
        assert.deepEqual(res, testdata.args, 'event data does not match')
        return q.disconnect()
      })
      q.cmd('servernotifyregister', {
        event: 'server'
      }, err => err ? next(err) : true)
    })

    it('should ignore responses without requests', next => {
      let fake = new FakeServer({})
      let testdata = {
        bools: ['error'],
        args: {
          id: 0,
          msg: 'ok'
        }
      }
      fake.addTarget('servernotifyregister', testdata)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.cmd('servernotifyregister', {
        event: 'server'
      }, err => err ? next(err) : q.disconnect())
    })

    it('should handle errors', next => {
      let fake = new FakeServer({})
      let testdata = {
        bools: ['error'],
        args: {
          id: 2,
          msg: 'fatal'
        }
      }
      fake.addTarget('servernotifyregister', testdata)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.cmd('servernotifyregister', {
        event: 'server'
      }, err => expect(() => {
        throw err
      }).to.throw('ServerError: fatal') && q.disconnect())
    })

    it('should handle flood ban', next => {
      const stream = {
        sink: pull.drain(() => {}),
        source: pull.values(['error id=3329 msg=connection\\sfailed,\\syou\\sare\\sbanned extra_msg=you\\smay\\sretry\\sin\\s600\\sseconds'])
      }
      const q = new Query()
      q.once('connect:done', err => expect(() => {
        throw err
      }).to.throw('FloodError: connection failed, you are banned') && next())
      q.connectPullStream(stream)
    })
  })
})

describe('pull', () => {
  describe('packer', () => {
    it('should pack the login command right', () => {
      pull(
        pull.values([{
          cmd: 'login',
          args: {
            bools: ['user', 'pw']
          }
        }]),
        mod.packer(),
        pull.drain(res => expect(res).to.equal('login user pw'))
      )
    })
    it('should pack an array', () => {
      pull(
        pull.values([{
          cmd: 'test',
          args: [{
            bools: ['t1'],
            args: {
              test: '1'
            }
          }, {
            bools: ['t2'],
            args: {
              test: '2'
            }
          }]
        }]),
        mod.packer(),
        pull.drain(res => expect(res).to.equal('test t1 test=1|t2 test=2'))
      )
    })
  })

  describe('parser', () => {
    it('should ignore the first messages and only parse the last one', next => {
      pull(
        pull.values(['TS3', 'Welcome', 'test']),
        mod.parser(),
        pull.drain(res => assert.deepEqual(res, [{
          test: true
        }]) || next())
      )
    })
    it('should parse arrays', next => {
      pull(
        pull.values(['TS3', 'Welcome', 'hello|world']),
        mod.parser(),
        pull.drain(res => assert.deepEqual(res, [{
          hello: true
        }, {
          world: true
        }]) || next())
      )
    })
  })
})


test/index.js
'use strict'

// const sinon = require("sinon")
const chai = require('chai')
chai.use(require('sinon-chai'))
chai.should()
const expect = chai.expect
const assert = require('assert')

const Query = require('..')
const mod = require('../pull')
const pull = require('pull-stream')
const FakeServer = require('./fake-ts3')

describe('query', () => {
  describe('connection', () => {
    it('should connect and login', next => {
      let fake = new FakeServer({
        login: ['serveradmin', 'pw123']
      })
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.login('serveradmin', 'pw123', err => err ? next(err) : true)
      q.disconnect()
    })

    it('should login, set the nickname and get the clientlist', next => {
      let fake = new FakeServer({
        login: ['serveradmin', 'pw123']
      })
      let cllist = [{
        args: {
          clid: 1,
          cid: 1,
          client_database_id: 1,
          client_nickname: 'Some Guy'
        }
      },
        {
          args: {
            clid: 2,
            cid: 2,
            client_database_id: 2,
            client_nickname: 'Another Guy'
          }
        }
      ]
      fake.addTarget('clientupdate')
      fake.addTarget('use')
      fake.addTarget('clientlist', cllist)
      fake.addTarget('clientlist', cllist)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.login('serveradmin', 'pw123', err => err ? next(err) : true)
      q.cmd('clientupdate', {
        client_nickname: 'test'
      }, err => err ? next(err) : true)
      q.cmd('use', 1, err => err ? next(err) : true)
      q.cmd('clientlist', (err, res) => err ? next(err) : assert.deepEqual(cllist.map(c => c.args), res, 'list differs'))
      q.list('clientlist', ['-uid'], (err, res) => err ? next(err) : assert.deepEqual(cllist.map(c => c.args), res, 'list differs'))
      q.disconnect()
    })

    it('should handle notify events', next => {
      let fake = new FakeServer({})
      let testdata = {
        bools: ['notifyclientleftview'],
        args: {
          client_nickname: 'Another guy'
        }
      }
      fake.addTarget('servernotifyregister', testdata)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.on('clientleftview', res => {
        assert.deepEqual(res, testdata.args, 'event data does not match')
        return q.disconnect()
      })
      q.cmd('servernotifyregister', {
        event: 'server'
      }, err => err ? next(err) : true)
    })

    it('should ignore responses without requests', next => {
      let fake = new FakeServer({})
      let testdata = {
        bools: ['error'],
        args: {
          id: 0,
          msg: 'ok'
        }
      }
      fake.addTarget('servernotifyregister', testdata)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.cmd('servernotifyregister', {
        event: 'server'
      }, err => err ? next(err) : q.disconnect())
    })

    it('should handle errors', next => {
      let fake = new FakeServer({})
      let testdata = {
        bools: ['error'],
        args: {
          id: 2,
          msg: 'fatal'
        }
      }
      fake.addTarget('servernotifyregister', testdata)
      const q = new Query()
      q.connectPullStream(fake.createStream(next))
      q.cmd('servernotifyregister', {
        event: 'server'
      }, err => expect(() => {
        throw err
      }).to.throw('ServerError: fatal') && q.disconnect())
    })

    it('should handle flood ban', next => {
      const stream = {
        sink: pull.drain(() => {}),
        source: pull.values(['error id=3329 msg=connection\\sfailed,\\syou\\sare\\sbanned extra_msg=you\\smay\\sretry\\sin\\s600\\sseconds'])
      }
      const q = new Query()
      q.once('connect:done', err => expect(() => {
        throw err
      }).to.throw('FloodError: connection failed, you are banned') && next())
      q.connectPullStream(stream)
    })
  })
})

describe('pull', () => {
  describe('packer', () => {
    it('should pack the login command right', () => {
      pull(
        pull.values([{
          cmd: 'login',
          args: {
            bools: ['user', 'pw']
          }
        }]),
        mod.packer(),
        pull.drain(res => expect(res).to.equal('login user pw'))
      )
    })
    it('should pack an array', () => {
      pull(
        pull.values([{
          cmd: 'test',
          args: [{
            bools: ['t1'],
            args: {
              test: '1'
            }
          }, {
            bools: ['t2'],
            args: {
              test: '2'
            }
          }]
        }]),
        mod.packer(),
        pull.drain(res => expect(res).to.equal('test t1 test=1|t2 test=2'))
      )
    })
  })

  describe('parser', () => {
    it('should ignore the first messages and only parse the last one', next => {
      pull(
        pull.values(['TS3', 'Welcome', 'test']),
        mod.parser(),
        pull.drain(res => assert.deepEqual(res, [{
          test: true
        }]) || next())
      )
    })
    it('should parse arrays', next => {
      pull(
        pull.values(['TS3', 'Welcome', 'hello|world']),
        mod.parser(),
        pull.drain(res => assert.deepEqual(res, [{
          hello: true
        }, {
          world: true
        }]) || next())
      )
    })
  })
})


test/fake-ts3.js
'use strict'

const Pushable = require('pull-pushable')
const pull = require('pull-stream')
const mod = require('../src/pull')
const Pair = require('pull-pair')
const util = require('util')

const assert = require('assert')

function FakeServer (opt) {
  const self = this

  self.assertOk = () => {
    // throw if something isn't like it should be
    if (targets.length) {
      throw new Error('Targets left: ' + util.inspect(targets, { depth: null }))
    }
  }

  let targets = []

  self.addTarget = (cmd, data) =>
    targets.push({
      cmd,
      data
    })

  self.createStream = _cb => {
    let client = {}
    const socketToClient = Pair()
    const socketToServer = Pair()

    assert(_cb, 'cb is undefined')

    let first = true

    let out = Pushable()

    pull(
      socketToServer.source,
      mod.splitCRLF(),
      mod.parserServer(),
      pull.drain(data => {
        const cb = (err, data) => {
          if (err) {
            out.end(err)
          } else if (data) {
            data.forEach(data => out.push(data))
          }
        }

        function assertError (err) {
          _cb(new Error('TestError: ' + err))
          return cb(true) // eslint-disable-line standard/no-callback-literal
        }

        function ok (res) {
          if (res) {
            cb(null, [res, [{
              bools: ['error'],
              args: {
                id: 0,
                msg: 'ok'
              }
            }]])
          } else {
            cb(null, [
              [{
                bools: ['error'],
                args: {
                  id: 0,
                  msg: 'ok'
                }
              }]
            ])
          }
        }

        function err (id, msg) {
          return cb(null, [
            [{
              bools: ['error'],
              args: {
                id,
                msg
              }
            }]
          ])
        }

        if (opt.login && !client.auth) {
          if (data.cmd !== 'login') return assertError('Did not login')
          if (data.data[0][opt.login[0]] && data.data[0][opt.login[1]]) {
            client.auth = true
            ok()
          } else {
            client.auth = false
            err(512, 'wrong username or password')
          }
          return
        }
        switch (data.cmd) {
          case 'quit':
            try {
              self.assertOk()
            } catch (e) {
              return _cb(e)
            }
            _cb()
            return cb(true) // eslint-disable-line standard/no-callback-literal
          default: {
            if (targets.length) {
              const t = targets.shift()
              if (t.cmd !== data.cmd) return assertError(t.cmd + ' was supposed to be called but ' + data.cmd + ' got called')
              return ok(t.data)
            } else return assertError(data.cmd + " was called but it wasn't planned to be called")
          }
        }
      }, err => {
        out.end(err)
        try {
          self.assertOk()
        } catch (e) {
          return _cb(e)
        }
        _cb()
      })
    )

    pull(
      out,
      mod.packerServer(),
      pull.map(data => {
        if (first) {
          first = false
          return 'TS3\n\rWelcome\n\r' + data
        }
        return data
      }),
      mod.addCRLF(),
      socketToClient.sink
    )
    return {
      source: socketToClient.source,
      sink: socketToServer.sink
    }
  }
}

module.exports = FakeServer


src/pull.js
'use strict'

/* eslint-env mocha */

const Pushable = require('pull-pushable')

const pull = require('pull-stream')
const _pull = module.exports

/* const map = {
  "\\s": " ",
  "\\p": "|",
  "\\n": "\n",
  "\\f": "\f",
  "\\r": "\r",
  "\\t": "\t",
  "\\v": "\v",
  "\\\/": "\/",
  "\\\\": "\\"
} */

function escaper (s, re) {
  s = String(s)
  re.forEach(r => (s = s.replace(r[0], r[1])))
  return s
}

let decodeRegex = [
  [/\\s/g, ' '],
  [/\\p/g, '|'],
  [/\\n/g, '\n'],
  [/\\f/g, '\f'],
  [/\\r/g, '\r'],
  [/\\t/g, '\t'],
  [/\\v/g, '\v'],
  [/\\\//g, '\/'], // eslint-disable-line no-useless-escape
  [/\\\\/g, '\\']
]
let encodeRegex = [
  [/\\/g, '\\\\'],
  [/\//g, '\\/'],
  [/\|/g, '\\p'],
  [/\n/g, '\\n'],
  [/\r/g, '\\r'],
  [/\t/g, '\\t'],
  [/\v/g, '\\v'],
  [/\f/g, '\\f'],
  [/ /g, '\\s']
]
/* for (var decoded in map) {
  decodeRegex.push([new RegExp(map[decoded].replace(/\\/g,"\\\\"), "g"), decoded]) // /encoded/g -> decoded
  encodeRegex.push([new RegExp(decoded.replace(/\\/g,"\\\\"), "g"), map[decoded]]) // /decoded/g -> encoded
} */

_pull.splitCRLF = () => {
  let data = ''
  let source = Pushable()

  return {
    source,
    sink: pull.drain(chunk => {
      data += chunk || ''
      let lines = data.split('\n\r')
      data = lines.pop()
      lines.forEach(l => source.push(l))
    }, err => {
      if (data) source.push(data)
      data = null
      source.end(err)
    })
  }
}

_pull.parser = function Parser () {
  let loc = -2
  let source = Pushable()

  return {
    source,
    sink: pull.drain(line => {
      if (loc) {
        loc++
        if (!line.startsWith('error')) return
      }
      let res = line.split('|').map(line => {
        line = line.split(' ')
        let res = {}
        line.forEach(line => {
          const eqPos = line.indexOf('=')
          if (eqPos !== -1) {
            const key = escaper(line.substr(0, eqPos), decodeRegex)
            let val = escaper(line.substr(eqPos + 1), decodeRegex)
            if (parseInt(val, 10).toString() === val) val = parseInt(val, 10)
            res[key] = val
          } else {
            res[line] = true
          }
        })
        return res
      })
      source.push(res)
    }, e => source.end(e))
  }
}

_pull.parserServer = function Parser () {
  return pull.map(line => {
    let cmd = line.split(' ').shift()
    let res = line.split(' ').slice(1).join(' ').split('|').map(line => {
      line = line.split(' ')
      let res = {}
      line.forEach(line => {
        const eqPos = line.indexOf('=')
        if (eqPos !== -1) {
          const key = escaper(line.substr(0, eqPos), decodeRegex)
          let val = escaper(line.substr(eqPos + 1), decodeRegex)
          if (parseInt(val, 10).toString() === val) val = parseInt(val, 10)
          res[key] = val
        } else {
          res[line] = true
        }
      })
      return res
    })
    return {
      cmd,
      data: res
    }
  })
}

_pull.packer = () => pull.map(data => {
  if (!Array.isArray(data.args)) data.args = [data.args]
  return data.cmd + ' ' + data.args.map(data => [(data.bools || []).map(d => escaper(d, encodeRegex)).join(' '),
    Object.keys(data.args || {}).map(k => escaper(k, encodeRegex) + '=' + escaper(data.args[k], encodeRegex)).join(' ')
  ].filter(e => e.length).join(' ')).join('|')
})

_pull.packerServer = () => pull.map(data => {
  if (!Array.isArray(data)) data = [data]
  return data.map(data => [(data.bools || []).map(d => escaper(d, encodeRegex)).join(' '),
    Object.keys(data.args || {}).map(k => escaper(k, encodeRegex) + '=' + escaper(data.args[k], encodeRegex)).join(' ')
  ].filter(e => e.length).join(' ')).join('|')
})

_pull.addCRLF = () => pull.map(d => d + '\n\r')


src/index.js
'use strict'

/* eslint-env mocha */

const pull = require('pull-stream')
const net = require('net')
const toPull = require('stream-to-pull-stream')
const mods = require('./pull')
const debug = require('debug')
const log = debug('teamspeak-query-client')
const uuid = require('uuid')

const EE = require('events').EventEmitter

function TeamSpeakQueryClient (opt) {
  const self = this

  EE.call(self)

  if (!opt) opt = {}
  opt.host = opt.host || 'localhost'
  opt.port = opt.port || 10011

  function queue (cmd, args, cb) {
    /*
    data={
      bool: [] //will be sent as is (encoded)
      args: {} //will be sent as encoded key=val
    }
    data can also be an array that will be encoded and joined with |
    */
    log('queued', cmd, cmd === 'login' ? '*' : args)
    que.push({
      data: {
        cmd,
        args
      },
      cb,
      id: uuid()
    })
    self.emit('que:add')
  }

  let que = []

  let ended = false

  let cur = null

  function handler () {
    return {
      sink: function (read) {
        read(null, function next (end, res) {
          if (end) {
            self.connected = false
            self.emit('disconnect')
            return (ended = end)
          }
          if (ended) return read(ended = end)
          if (res.length === 1 && Object.keys(res[0])[0].startsWith('notify')) {
            // notify event (see "servernotifyregister" command)
            let ev = Object.keys(res[0])[0].substr('notify'.length)
            delete res[0]['notify' + ev]
            log('notify', ev, res[0])
            self.emit(ev, res[0])
            self.emit('notify', ev, res[0])
            return read(null, next)
          }
          if (res.length === 1 && res[0].error && (res[0].id === 3352 || res[0].id === 3329)) { // you are flood-banned
            const e = new Error('FloodError: ' + res[0].msg)
            e.id = res[0].id
            self.emit('connect:done', e)
            return
          }
          if (!cur) {
            log('got response with no request', res)
            return read(end, next)
          }
          log('got response', res)
          if (res.length === 1 && res[0].error) {
            // error response, done with request
            let err = res[0]
            if (!err.id) cur.err = null
            else {
              cur.err = new Error('ServerError: ' + err.msg)
              cur.err.id = err.id
            }
            cur.cb(cur.err, cur.resp)
            cur = null
            self.emit('que:done')
          } else { // result for command
            cur.resp = res
          }
          return read(null, next)
        })
      },
      source: function (end, cb) {
        if (end) {
          ended = true
          return cb(end)
        }

        function doSend () {
          cur = que.shift()
          log('sending', cur.data.cmd === 'login' ? '*' : cur.data)
          self.emit('que:do')
          return cb(null, cur.data)
        }

        function send () {
          if (!cur) {
            if (que.length) doSend()
            else self.once('que:add', doSend)
          } else self.once('que:done', send)
        }

        send()
      }
    }
  }

  function streamConnect (stream) {
    self.connected = true
    let first = true
    self.stream = pull( // glue it together
      stream.source,
      mods.splitCRLF(),
      pull.map(d => {
        if (first) {
          setTimeout(() => self.emit('connect:done'), 100)
          first = false
        }
        log('raw_in', d)
        return d
      }),
      mods.parser(),
      handler(),
      mods.packer(),
      pull.map(d => {
        log('raw_out', d.startsWith('login') ? 'login *** ***' : d)
        return d
      }),
      mods.addCRLF(),
      stream.sink
    )
    self.emit('connected')
  }

  self.connect = cb => {
    const conn = net.connect(opt.port, opt.host, err => {
      if (err) return cb(err)

      const stream = toPull.duplex(conn)
      streamConnect(stream)
      log('connected', opt)
      self.once('connect:done', cb)
    })
  }

  self.disconnect = cb => {
    if (!self.connected) return cb()
    queue('quit', {}, cb)
  }

  self.connectStream = stream => streamConnect(toPull.duplex(stream))
  self.connectPullStream = stream => streamConnect(stream)

  self.login = (user, pw, cb) => {
    queue('login', {
      bools: [user, pw]
    }, cb)
  }

  self.cmd = (cmd, args, bools, cb) => {
    if (typeof args === 'function') {
      cb = args
      args = {}
      bools = []
    }
    if (typeof bools === 'function') {
      cb = bools
      bools = null
    }
    if (Array.isArray(args) && !bools) {
      bools = args
      args = {}
    }
    if (!bools) bools = []
    if (!Array.isArray(bools)) bools = [bools]
    if (typeof args === 'number' || typeof args === 'string') {
      bools.push(args)
      args = {}
    }
    let res
    if (Array.isArray(args)) {
      res = args.map(a => {
        if (a.args) {
          return a
        } else {
          a = {
            args: a,
            bools: []
          }
        }
      })
    } else {
      res = {
        args,
        bools: bools
      }
    }
    queue(cmd, res, cb)
  }

  self.list = (cmd, args, bools, cb) => {
    self.cmd(cmd, args, bools, (err, res) => {
      if (err) return cb(err)
      if (!res) res = []
      if (!Array.isArray(res)) res = [res]
      cb(null, res)
    })
  }
}

require('util').inherits(TeamSpeakQueryClient, EE)

module.exports = TeamSpeakQueryClient
